### ISSUE TITLE
bazel-* files on the top level directory are not linked in the execroot, even when they are explicitely defined.

### ISSUE THREAD

**Issue by davidzchen** [bazelbuild/bazel#2909]
### Description of the problem / feature request / question:

When running `bazel build //:jekyll-tree` in the following workspace (https://github.com/davidzchen/bazel-website/tree/third-party), the build fails with mysterious errors about files not being found. The first time it is run, it complains that `bazel-release.pub.gpg` cannot be found, but the file exists and should have been picked up by the file glob. During subsequent runs, it complains that `//:bazel-out/_tmp/action_outs/stderr-17` being missing (though the number after `stderr-` changes for each subsequent run.

### If possible, provide a minimal example to reproduce the problem:

1. Clone the following branch of bazelbuild/bazel-website: https://github.com/davidzchen/bazel-website/tree/third-party
2. Run `bazel build //:jekyll-tree`

### Environment info

* Operating System: Ubuntu Trusty and macOS Sierra

* Bazel version (output of `bazel info release`): Confirmed on 0.4.3 and current HEAD (`0.4.5-2017-04-28 (@b8449f958)`)

###  Have you found anything relevant by searching the web?

Nothing so far.

### Anything else, information or logs or outputs that would be helpful?

With sandboxing:

```
❯❯❯ bazel build :jekyll-tree
INFO: Analysed target //:jekyll-tree (13 packages loaded).
INFO: Found 1 target...
ERROR: /Volumes/Ocean/Projects/bazelbuild/bazel-website/BUILD:16:1: error executing shell command: 'bazel-out/host/bin/external/bazel_tools/tools/build_defs/pkg/build_tar --flagfile=bazel-out/local-fastbuild/bin/jekyll-files.args' failed: I/O exception during sandboxed execution: /private/var/tmp/_bazel_dzc/796c6b92b14d67d52278cb3fde5a83a6/execroot/bazel-website/bazel-release.pub.gpg (No such file or directory)
Target //:jekyll-tree failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 1.551s, Critical Path: 0.12s
FAILED: Build did NOT complete successfully
❯❯❯ bazel build :jekyll-tree
INFO: Analysed target //:jekyll-tree (1 packages loaded).
INFO: Found 1 target...
ERROR: missing input file '//:bazel-out/_tmp/action_outs/stderr-15'
ERROR: /Volumes/Ocean/Projects/bazelbuild/bazel-website/BUILD:22:1: error executing shell command: 'bazel-out/host/bin/external/bazel_tools/tools/build_defs/pkg/build_tar --flagfile=bazel-out/local-fastbuild/bin/bootstrap-css.args' failed: Process terminated by signal 15 [sandboxed]
Use --strategy=PackageTar=standalone to disable sandboxing for the failing actions.
ERROR: /Volumes/Ocean/Projects/bazelbuild/bazel-website/BUILD:16:1: //:jekyll-files: missing input file '//:bazel-out/_tmp/action_outs/stderr-15'
Target //:jekyll-tree failed to build
Use --verbose_failures to see the command lines of failed build steps.
ERROR: /Volumes/Ocean/Projects/bazelbuild/bazel-website/BUILD:16:1 1 input file(s) do not exist
INFO: Elapsed time: 0.342s, Critical Path: 0.13s
FAILED: Build did NOT complete successfully
❯❯❯ bazel build :jekyll-tree
INFO: Analysed target //:jekyll-tree (1 packages loaded).
INFO: Found 1 target...
ERROR: missing input file '//:bazel-out/_tmp/action_outs/stderr-3'
ERROR: /Volumes/Ocean/Projects/bazelbuild/bazel-website/BUILD:36:1: error executing shell command: 'bazel-out/host/bin/external/bazel_tools/tools/build_defs/pkg/build_tar --flagfile=bazel-out/local-fastbuild/bin/font-awesome-css.args' failed: Process terminated by signal 15 [sandboxed]
Use --strategy=PackageTar=standalone to disable sandboxing for the failing actions.
ERROR: /Volumes/Ocean/Projects/bazelbuild/bazel-website/BUILD:16:1: //:jekyll-files: missing input file '//:bazel-out/_tmp/action_outs/stderr-3'
Target //:jekyll-tree failed to build
Use --verbose_failures to see the command lines of failed build steps.
ERROR: /Volumes/Ocean/Projects/bazelbuild/bazel-website/BUILD:16:1 1 input file(s) do not exist
INFO: Elapsed time: 0.288s, Critical Path: 0.10s
FAILED: Build did NOT complete successfully
❯❯❯ bazel build :jekyll-tree
INFO: Analysed target //:jekyll-tree (0 packages loaded).
INFO: Found 1 target...
ERROR: missing input file '//:bazel-out/_tmp/action_outs/stderr-6'
ERROR: /Volumes/Ocean/Projects/bazelbuild/bazel-website/BUILD:50:1: error executing shell command: 'bazel-out/host/bin/external/bazel_tools/tools/build_defs/pkg/build_tar --flagfile=bazel-out/local-fastbuild/bin/bootstrap-js.args' failed: Process terminated by signal 15 [sandboxed]
Use --strategy=PackageTar=standalone to disable sandboxing for the failing actions.
ERROR: /Volumes/Ocean/Projects/bazelbuild/bazel-website/BUILD:16:1: //:jekyll-files: missing input file '//:bazel-out/_tmp/action_outs/stderr-6'
Target //:jekyll-tree failed to build
Use --verbose_failures to see the command lines of failed build steps.
ERROR: /Volumes/Ocean/Projects/bazelbuild/bazel-website/BUILD:16:1 1 input file(s) do not exist
INFO: Elapsed time: 0.207s, Critical Path: 0.09s
FAILED: Build did NOT complete successfully
❯❯❯ bazel build :jekyll-tree
INFO: Analysed target //:jekyll-tree (0 packages loaded).
INFO: Found 1 target...
ERROR: missing input file '//:bazel-out/_tmp/action_outs/stderr-5'
ERROR: /Volumes/Ocean/Projects/bazelbuild/bazel-website/BUILD:36:1: error executing shell command: 'bazel-out/host/bin/external/bazel_tools/tools/build_defs/pkg/build_tar --flagfile=bazel-out/local-fastbuild/bin/font-awesome-css.args' failed: Process terminated by signal 15 [sandboxed]
Use --strategy=PackageTar=standalone to disable sandboxing for the failing actions.
ERROR: /Volumes/Ocean/Projects/bazelbuild/bazel-website/BUILD:16:1: //:jekyll-files: missing input file '//:bazel-out/_tmp/action_outs/stderr-5'
Target //:jekyll-tree failed to build
Use --verbose_failures to see the command lines of failed build steps.
ERROR: /Volumes/Ocean/Projects/bazelbuild/bazel-website/BUILD:16:1 1 input file(s) do not exist
INFO: Elapsed time: 0.187s, Critical Path: 0.08s
FAILED: Build did NOT complete successfully
```

Running without sandboxing gives the following stack trace but on subsequent runs, bazel complains about missing the stderr file under `bazel-out` again.

```
❯❯❯ bazel build :jekyll-tree --strategy=PackageTar=standalone
INFO: Analysed target //:jekyll-tree (13 packages loaded).
INFO: Found 1 target...
ERROR: /Volumes/Ocean/Projects/bazelbuild/bazel-website/BUILD:16:1: error executing shell command: 'bazel-out/host/bin/external/bazel_tools/tools/build_defs/pkg/build_tar --flagfile=bazel-out/local-fastbuild/bin/jekyll-files.args' failed: bash failed: error executing command /bin/bash -c 'bazel-out/host/bin/external/bazel_tools/tools/build_defs/pkg/build_tar --flagfile=bazel-out/local-fastbuild/bin/jekyll-files.args': com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 1
Traceback (most recent call last):
  File "/private/var/tmp/_bazel_dzc/796c6b92b14d67d52278cb3fde5a83a6/execroot/bazel-website/bazel-out/host/bin/external/bazel_tools/tools/build_defs/pkg/build_tar.runfiles/build_bazel_website/../bazel_tools/tools/build_defs/pkg/build_tar.py", line 238, in <module>
    main(FLAGS(sys.argv))
  File "/private/var/tmp/_bazel_dzc/796c6b92b14d67d52278cb3fde5a83a6/execroot/bazel-website/bazel-out/host/bin/external/bazel_tools/tools/build_defs/pkg/build_tar.runfiles/build_bazel_website/../bazel_tools/tools/build_defs/pkg/build_tar.py", line 227, in main
    output.add_file(inf, tof, mode, ids, names)
  File "/private/var/tmp/_bazel_dzc/796c6b92b14d67d52278cb3fde5a83a6/execroot/bazel-website/bazel-out/host/bin/external/bazel_tools/tools/build_defs/pkg/build_tar.runfiles/build_bazel_website/../bazel_tools/tools/build_defs/pkg/build_tar.py", line 121, in add_file
    gname=names[1])
  File "/private/var/tmp/_bazel_dzc/796c6b92b14d67d52278cb3fde5a83a6/execroot/bazel-website/bazel-out/host/bin/external/bazel_tools/tools/build_defs/pkg/build_tar.runfiles/bazel_tools/tools/build_defs/pkg/archive.py", line 269, in add_file
    with open(file_content, 'rb') as f:
IOError: [Errno 2] No such file or directory: 'bazel-release.pub.gpg'
Target //:jekyll-tree failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 3.918s, Critical Path: 0.52s
FAILED: Build did NOT complete successfully
❯❯❯ bazel build :jekyll-tree --strategy=PackageTar=standalone
INFO: Analysed target //:jekyll-tree (1 packages loaded).
INFO: Found 1 target...
ERROR: missing input file '//:bazel-out/_tmp/action_outs/stderr-17'
ERROR: /Volumes/Ocean/Projects/bazelbuild/bazel-website/BUILD:16:1: //:jekyll-files: missing input file '//:bazel-out/_tmp/action_outs/stderr-17'
Target //:jekyll-tree failed to build
Use --verbose_failures to see the command lines of failed build steps.
ERROR: /Volumes/Ocean/Projects/bazelbuild/bazel-website/BUILD:16:1 1 input file(s) do not exist
INFO: Elapsed time: 0.182s, Critical Path: 0.00s
FAILED: Build did NOT complete successfully
```

**Comment by davidzchen** [comment]
It looks like the problem was that name of the file in question, `bazel-release.pub.gpg`, starts with `bazel-`. Renaming this file to `bazel_release.pub.gpg` and adding `bazel-*/**` to the `exclude` list for the glob solved this problem. This is worth debugging further because it seems that any regular file with a name that begins with the string `"bazel-"` would cause `pkg_tar` to raise this error. For instance:

```
❯❯❯ touch bazel-foo.md
❯❯❯ bazel build :jekyll-tree
INFO: Analysed target //:jekyll-tree (1 packages loaded).
INFO: Found 1 target...
ERROR: /Volumes/Ocean/Projects/bazelbuild/bazel-website/BUILD:18:1: error executing shell command: 'bazel-out/host/bin/external/bazel_tools/tools/build_defs/pkg/build_tar --flagfile=bazel-out/darwin_x86_64-fastbuild/bin/jekyll-files.args' failed: I/O exception during sandboxed execution: /private/var/tmp/_bazel_dzc/796c6b92b14d67d52278cb3fde5a83a6/execroot/bazel-website/bazel-foo.md (No such file or directory)
Target //:jekyll-tree failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 0.601s, Critical Path: 0.12s
FAILED: Build did NOT complete successfully
```

**Comment by damienmg** [comment]
Can you share the content of the offending build files? It seems to me that you are globbing the convenience symlinks of Bazel.

**Comment by davidzchen** [comment]
Here is the BUILD file in my `bazel-website` branch where I am working on this: https://github.com/davidzchen/bazel-website/blob/third-party/BUILD.

I have added the convenience symlinks to the `exclude` list for the glob, but adding any file that begins with `bazel-` would still cause this error to occur.

**Comment by damienmg** [comment]
Sorry what do you meant by "but adding any file that begins with bazel- would
still cause this error to occur"? The problem is the '**' globs bazel-*
which include the convenience symlink.

**Comment by davidzchen** [comment]
I mean that if I rename `bazel_release.pub.gpg` back to `bazel-release.pub.gpg` or simply `touch bazel-foo.md`, the same error reproduces. Would the `bazel-*/**` in the `exclude` list exclude the convenience symlinks, even though the include glob pattern is `**`?

**Comment by damienmg** [comment]
Repro case:

```
[dmarting@dmarting-macbookpro] /tmp/test$ ls
[dmarting@dmarting-macbookpro] /tmp/test$ touch WORKSPACE
[dmarting@dmarting-macbookpro] /tmp/test$ touch bazel-release
[dmarting@dmarting-macbookpro] /tmp/test$ cat <<EOF >BUILD
> genrule(name = "test", outs = ["output.txt"], cmd = "ls -R | tee $@", srcs = ["bazel-release"])
> EOF
[dmarting@dmarting-macbookpro] /tmp/test$ bazel build :tet
........^C
[dmarting@dmarting-macbookpro] /tmp/test$ bazel build :test
INFO: Analysed target //:test.
INFO: Found 1 target...
ERROR: /private/tmp/test/BUILD:1:1: Executing genrule //:test failed: Could not prepare sandbox directory: /private/var/tmp/_bazel_dmarting/1063c9c2120b66f3130a9f592ab85144/execroot/test/bazel-release (No such file or directory)
Target //:test failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 2.691s, Critical Path: 0.01s
FAILED: Build did NOT complete successfully
```

**Comment by meisterT** [comment]
This seems to be fixed with Bazel 3.1
